stl/inc/ranges

```diff
@@ -6837,6 +6837,664 @@ namespace ranges {
+                            _STD get<_CurrIndex>(_Lhs._M_Current) - _STD get<_CurrIndex>(_Rhs._M_Current)};
+                        const auto _Curr_abs_difference_value = difference_type{_RANGES distance(
+                            _STD get<_CurrIndex>(_Lhs._M_Current), _STD get<_CurrIndex>(_Rhs._M_Current))};
+
+                        if (_Curr_abs_difference_value < _Smallest_abs_difference_value) {
+                            _Smallest_difference_value     = _Curr_difference_value;
+                            _Smallest_abs_difference_value = _Curr_abs_difference_value;
+                        }
+
+                        static constexpr auto _Next_index = (_CurrIndex + 1);
+                        self.operator()<_Next_index>();
+                    }
+                };
+
+                // Start from index 1, since we already evaluated 0.
+                _Evaluate_Distances_Lambda.operator()<1>();
+
+                return _Smallest_difference_value;
+            }
+        }
+
+        _NODISCARD_FRIEND constexpr auto iter_move(const _Iterator& _Itr) noexcept(
+            (noexcept(_RANGES iter_move(_STD declval<const _RANGES iterator_t<_Maybe_const<_IsConst, _ViewTypes>>&>()))
+                && ...)
+            && (_STD is_nothrow_move_constructible_v<
+                    _RANGES range_rvalue_reference_t<_Maybe_const<_IsConst, _ViewTypes>>> && ...)) {
+            return _Zip::_Tuple_transform(_RANGES iter_move, _Itr._M_Current);
+        }
+
+        friend constexpr void iter_swap(const _Iterator& _Lhs, const _Iterator& _Rhs) noexcept((
+            noexcept(_RANGES iter_swap(_STD declval<_RANGES iterator_t<_Maybe_const<_IsConst, _ViewTypes>>>(),
+                _STD declval<_RANGES iterator_t<_Maybe_const<_IsConst, _ViewTypes>>>()))
+            && ...)) requires(_STD indirectly_swappable<_RANGES iterator_t<_Maybe_const<_IsConst, _ViewTypes>>>&&...) {
+            const auto _Perform_swap_lambda = [&_Lhs, &_Rhs]<_STD size_t _CurrIndex>(this const auto& self) {
+                if constexpr (_CurrIndex != sizeof...(_ViewTypes)) {
+                    _RANGES iter_swap(_STD get<_CurrIndex>(_Lhs._M_Current), _STD get<_CurrIndex>(_Rhs._M_Current));
+
+                    static constexpr auto _Next_index = (_CurrIndex + 1);
+                    self.operator()<_Next_index>();
+                }
+            };
+
+            _Perform_swap_lambda.operator()<0>();
+        }
+
+    private:
+        _Iterator_tuple_type _M_Current;
+    };
+
+    template <_RANGES input_range... _ViewTypes>
+        requires _Zip::_Allow_views<_ViewTypes...>
+    template <bool _IsConst>
+    class zip_view<_ViewTypes...>::_Sentinel {
+    private:
+        friend class zip_view<_ViewTypes...>;
+
+        template <bool _SentinelConst, bool _IteratorConst, _RANGES input_range... _Views>
+        requires _Zip::_Allow_views<_Views...> _NODISCARD_FRIEND constexpr bool _Zip::_Iterator_sentinel_equality_impl(

```

Comment:
These friend declarations fail to compile. Can we turn these functions into members of iterator types, and make iterator a friend?

Reply:
Yes, I just noticed that this was causing compilation errors in Clang. (The MSVC accepts it, though.) I pushed a commit to change this.

Reply:
For those wondering, the actual reason for the compilation errors here is that according to the C++ specifications, attribute lists cannot appear within a friend function declaration unless said declaration is also a definition. So, Clang is technically right for rejecting the code. The MSVC and g++ both erroneously accept it by ignoring the attributes, but at least g++ issues a warning in this case when using `-Wall`. More information can be found at [this Stack Overflow post](https://stackoverflow.com/questions/52332203/friend-member-functions-and-attributes-gcc-vs-clang).